### PR TITLE
Adds goon health monitor implants

### PR DESCRIPTION
**Pull request by ghost**
Sprites are from goon, but the code is pretty much mine :D

:cl: More Robust Than You
add: RnD can now make a Health Monitor implant, which can allow you to see your own brute/burn/oxygen/toxin damage without a health analyzer! In addition, it will alert Medbay if you go into crit! (Thanks ninjanomnom for color math!)
/:cl:

### CODE REVIEW

**Comment by FantasticFwoosh** [comment]
Seems like clutter.

**Comment by MMMiracles** [comment]
'clutter' that only exists if you bother getting the implant in the first place, which means you actively want to see that information.

fuck i just replied to fwoosh

**Comment by praisenarsie** [comment]
@FantasticFwoosh you're clutter

**Comment by Cyberboss** [comment]
I'm not reviewing this till you test it

**Comment by MoreRobustThanYou** [comment]
I did test it. How else did I get those pics

**Comment by kevinz000** [comment]
fwoosh is clutter

**Comment by FantasticFwoosh** [comment]
> 'clutter' that only exists if you bother getting the implant in the first place, which means you actively want to see that information.
 
Just use a health scanner or a medical hud without putting a invasive implant inside yourself, **lmao** man just close your eyes and listen to your cardiac heartbeat for response of respiratory damage 🤣

**Comment by MoreRobustThanYou** [comment]
can we repoban fwoosh yet

**Comment by awesine** [comment]
>acknowledging that fwoosh even exists

**Comment by FantasticFwoosh** [comment]
The pot calling the kettle black.

**Comment by praisenarsie** [comment]
>acknowledging that awesine even exists

**Comment by Shapsy** [comment]
The seeing health part is a very neat addition, looks cool!

**Comment by Cyberboss** [comment]
I can tell by glancing at the code you didn't test it properly

**Comment by awesine** [comment]
@FantasticFwoosh 
well i'm a GRAY pot!

**Comment by MoreRobustThanYou** [comment]
@Cyberboss what do you mean?

**Comment by kevinz000** [comment]
good

**Comment by GunHog** [comment]
Very good addition :+1:

**Comment by Anonmare** [comment]
I'd replace the toxins sign with a biohazard, doesn't immediately scream "toxin" to me.

**Comment by MoreRobustThanYou** [comment]
@Anonmare i didn't sprite these, i just ripped the sprites from goonstation-2016.

**Comment by DoctoreMad** [comment]
~~Not necessary since a player can just look and see all other 3 damage types are covered~~ TGstation might actually be a bit too dumb for that, actually.

**Comment by MoreRobustThanYou** [comment]
@DoctoreMad yeah, tbh, goonstation players are usually smarter than tg players

**Comment by calzilla1** [comment]
Hahaha no

**Comment by awesine** [comment]
goonstation best station merge post-haste

**Comment by awesine** [comment]
soon /tg/station will just be goonstation

**Comment by awesine** [comment]
all stations?
goonstations

**Comment by awesine** [comment]
V.O.R.E?
G.O.O.N

**Comment by awesine** [comment]
/tg/?
/SA/

**Comment by awesine** [comment]
you wish @calzilla1

**Comment by MoreRobustThanYou** [comment]
calm down @awesine

**Comment by yorii** [comment]
let me get in on this spam train

**Comment by uraniummeltdown** [comment]
awesome, keep the goon sprites

**Comment by ExcessiveUseOfCobblestone** [comment]
Do people not realize this is an implant like tracking/chem/exile implants and not like Tool-arm, CNS, etc implants?

It's literally a instant, 1click to self and now you have epic harmalarm.

It's really only balanced by how quickly subdermal implants gets researched, which in MRTY's favor I don't think is a rushed tech (yet) but this is **REALLY REALLY good**.

**Comment by MoreRobustThanYou** [comment]
@ExcessiveUseOfCobblestone By the way, the harm alarm only activates once, unless you exit crit.

**Comment by ExcessiveUseOfCobblestone** [comment]
The concern is the balance, not spam although there are times where you can go in and out of crit.

It doesn't even delete after you trigger the harm alarm :cry:

**Comment by MoreRobustThanYou** [comment]
ofc a harm alarm means nothing if nobody bothers to retrieve the body

**Comment by Cyberboss** [comment]
>the harm alarm only activates once, unless you exit crit.

He's become so good at lying he actually believes himself

File: code/game/objects/items/implants/implant_misc.dm [hunk 1]
```diff
@@ -114,3 +114,40 @@
...
+		to_chat(C, "<span class='notice'>You feel more in-tune with your body.</span>")
+	return ..()
+
+/obj/item/implant/health_monitor/on_life(mob/living/carbon/source)
+	if(!triggered_in_crit && source.InCritical())
+		triggered_in_crit = TRUE
+		var/area/location = get_area(src)
+		internal_radio.talk_into(src, "Medical emergency! [source] is in critical condition at [location]!", "Medical", SPAN_ROBOT)
+	else
+		triggered_in_crit = FALSE
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/implants/implant_misc.dm]: Won't this spam if you're in crit?

Comment (MoreRobustThanYou) [reply code/game/objects/items/implants/implant_misc.dm]: No. The message is only triggered if triggered_in_crit is also false, and triggered_in_crit is set to true until you exit crit

Comment (ExcessiveUseOfCobblestone) [reply code/game/objects/items/implants/implant_misc.dm]: you have it as an AND not an OR so on next on_life it gets set right back to false then repeats over and over and ov...

**Comment by ShizCalev** [comment]
conflict.

File: code/_onclick/hud/hud.dm [hunk 2]
```diff
@@ -44,6 +44,8 @@
 	var/obj/screen/healthdoll
 	var/obj/screen/internals
 
+	var/list/obj/screen/adv_health = list()
```
Comment (Cyberboss) [inline code/_onclick/hud/hud.dm]: Make it a lazy list

File: code/_onclick/hud/hud.dm [hunk 3]
```diff
@@ -100,6 +102,11 @@
 			qdel(thing)
 		infodisplay.Cut()
 
+	if(adv_health.len)
+		for(var/thing in adv_health)
+			qdel(thing)
+		adv_health.Cut()
```
Comment (Cyberboss) [inline code/_onclick/hud/hud.dm]: This whole thing => QDEL_LIST(adv_health)

File: code/_onclick/hud/human.dm [hunk 4]
```diff
@@ -280,6 +280,12 @@
 	healths = new /obj/screen/healths()
 	infodisplay += healths
 
+	for(var/X in subtypesof(/obj/screen/adv_health))
+		var/obj/screen/adv_health/Y = new X()
+		Y.invisibility = INVISIBILITY_ABSTRACT
```
Comment (Cyberboss) [inline code/_onclick/hud/human.dm]: If this happens every time it should be a property of the object

File: code/game/objects/items/implants/implant_misc.dm [hunk 5]
```diff
@@ -114,3 +114,40 @@
...
+/obj/item/implant/health_monitor/Initialize()
+	. = ..()
+	internal_radio = new/obj/item/device/radio(src)
+	internal_radio.keyslot = new /obj/item/device/encryptionkey/headset_med
+	internal_radio.subspace_transmission = TRUE
+	internal_radio.listening = FALSE
+	internal_radio.broadcasting = FALSE
+	internal_radio.recalculateChannels()
+
+/obj/item/implant/health_monitor/removed(source, silent = 0, special = 0)
```
Comment (Cyberboss) [inline code/game/objects/items/implants/implant_misc.dm]: QDEL_NULL internal_radio in Destroy()

File: code/game/objects/items/implants/implant_misc.dm [hunk 6]
```diff
@@ -114,3 +114,40 @@
...
+	if(..())
```
Comment (Cyberboss) [inline code/game/objects/items/implants/implant_misc.dm]: . = ..()
if(.)

File: code/game/objects/items/implants/implant_misc.dm [hunk 7]
```diff
@@ -114,3 +114,40 @@
...
+		if(iscarbon(source))
```
Comment (Cyberboss) [inline code/game/objects/items/implants/implant_misc.dm]: Combine the two ifs

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/death.dm [hunk 8]
```diff
@@ -12,6 +12,11 @@
 	if(SSticker.mode)
 		SSticker.mode.check_win() //Calls the rounds wincheck, mainly for wizard, malf, and changeling now
 
+	if(adv_health_hud && hud_used && LAZYLEN(hud_used.adv_health))
+		for(var/obj/screen/adv_health/S in hud_used.adv_health)
```
Comment (Iamgoofball) [inline code/modules/mob/living/carbon/death.dm]: typeless loop

File: code/modules/mob/living/carbon/life.dm [hunk 9]
```diff
@@ -456,6 +461,39 @@
...
+///////////////////
+//ADVANCED HEALTH//
+///////////////////
+
+/mob/living/carbon/proc/handle_adv_hud()
+	if(adv_health_hud)
+		hud_used.healths.invisibility = INVISIBILITY_ABSTRACT
+	else
+		hud_used.healths.invisibility = 0
+	for(var/obj/screen/adv_health/S in hud_used.adv_health)
```
Comment (Iamgoofball) [inline code/modules/mob/living/carbon/life.dm]: typeless loop

**Comment by WJohn** [comment]
Sprites are good, except perhaps the fire notification, could perhaps be redrawn to look less like a tooth and be more chaotic like an actual fire.

**Comment by MoreRobustThanYou** [comment]
@WJohn i couldn't sprite my way out of a paper bag

**Comment by GunHog** [comment]
>@WJohn i couldn't sprite my way out of a paper bag

If I were coding it, I would just put "B" for "Brute", "F" for "Fire, "O" for "Oxygen", and "T" for "Toxin".

**Comment by praisenarsie** [comment]
@WJohn ARE YOU CRITICIZING GOON SPRITES?

@GunHog sounds lame

anyway, i'll make sprites for it (hopefully today if i feel like it)

**Comment by MoreRobustThanYou** [comment]
@praisenarsie Remember, each damage type has THREE sprites: The greyscale overlay, the box with the pale outline, and the box with the orange outline

**Comment by ZeWaka** [comment]
sprite rip makes me sad :(

**Comment by MoreRobustThanYou** [comment]
@ZeWaka At least goon gets the credit for the sprites

File: code/_onclick/hud/health_adv.dm [hunk 10]
```diff
@@ -0,0 +1,37 @@
+/obj/screen/adv_health
+	name = "YOU SHOULD NOT SEE THIS!"
+	icon = 'goon/icons/mob/mhealth.dmi'
```
Comment (ninjanomnom) [inline code/_onclick/hud/health_adv.dm]: get a non goon sprite or get permission from the artist so we don't add more stuff to the goon folder

**Review by ninjanomnom — DISMISSED** [review]

File: code/_onclick/hud/health_adv.dm (hunk 10 above)
Comment (MoreRobustThanYou) [reply code/_onclick/hud/health_adv.dm]: who tf is the artist

Comment (ExcessiveUseOfCobblestone) [reply code/_onclick/hud/health_adv.dm]: just make ur own sprites stupid cat

Comment (Iamgoofball) [reply code/_onclick/hud/health_adv.dm]: Did you guys realize Gannets posted in here

We can just ask him
@Gannets can we use your death laser while we work on our own

Comment (Iamgoofball) [reply code/_onclick/hud/health_adv.dm]: Wait shit wrong pr this isnt the laser beam

Comment (Gannets) [reply code/_onclick/hud/health_adv.dm]: Oh dear. Yeah, you're welcome to, pretty sure someone wandered in from Hippie to ask if they could use it a while back.

Also, not sure who drew the health monitor icons, I'd take a stab and assume Haine?

Comment (uraniummeltdown) [reply code/_onclick/hud/health_adv.dm]: There is literally nothing wrong with goonsprites, ignore the license autism Morty @MoreRobustThanYou

**Comment by AlexTheSysop42** [comment]
This is good but I think there should be an icon for overall health to get an idea of how dead you are if you have, say, a bunch of both brute and burn damage.

File: code/__DEFINES/maths.dm [hunk 11]
```diff
@@ -86,7 +86,7 @@
 // Performs a linear interpolation between a and b.
 // Note that amount=0 returns a, amount=1 returns b, and
 // amount=0.5 returns the mean of a and b.
-#define LERP(a, b, amount) (amount ? ((a) + ((b) - (a)) * (amount)) : ((a) + ((b) - (a)) * 0.5)
+#define LERP(a, b, amount) ( amount ? ((a) + ((b) - (a)) * (amount)) : a )
```
Comment (Cyberboss) [inline code/__DEFINES/maths.dm]: ???

File: code/_onclick/hud/health_adv.dm [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+	var/dmg_type = BRUTE
+
+/obj/screen/adv_health/Initialize()
+	. = ..()
+	color_overlay = mutable_appearance(icon, "[dmg_type]_overlay")
+	add_overlay(color_overlay)
+
+/obj/screen/adv_health/Destroy()
+	QDEL_NULL(color_overlay)
+
```
Comment (Cyberboss) [inline code/_onclick/hud/health_adv.dm]: return ..()

File: code/_onclick/hud/hud.dm [hunk 13]
```diff
@@ -63,6 +65,8 @@
 		plane_masters["[instance.plane]"] = instance
 		instance.backdrop(mymob)
 
+	LAZYINITLIST(adv_health)
```
Comment (Cyberboss) [inline code/_onclick/hud/hud.dm]: This isn't a lazy list if you init it in New()

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/objects/items/implants/implant_misc.dm [hunk 14]
```diff
@@ -114,3 +114,44 @@
...
+	internal_radio = new/obj/item/device/radio(src)
+	internal_radio.keyslot = new /obj/item/device/encryptionkey/headset_med
+	internal_radio.subspace_transmission = TRUE
+	internal_radio.listening = FALSE
+	internal_radio.broadcasting = FALSE
+	internal_radio.recalculateChannels()
+
+/obj/item/implant/health_monitor/Destroy()
+	. = ..()
+	QDEL_NULL(internal_radio)
```
Comment (Cyberboss) [inline code/game/objects/items/implants/implant_misc.dm]: reverse order and use return

File: code/game/objects/items/implants/implant_misc.dm [hunk 15]
```diff
@@ -114,3 +114,44 @@
...
+
+/obj/item/implant/health_monitor/removed(source, silent = 0, special = 0)
```
Comment (Cyberboss) [inline code/game/objects/items/implants/implant_misc.dm]: FALSE twice

File: code/game/objects/items/implants/implant_misc.dm [hunk 16]
```diff
@@ -114,3 +114,44 @@
...
+	. = ..()
+	if(. && iscarbon(source))
+		var/mob/living/carbon/C = source
+		C.adv_health_hud = FALSE
+		to_chat(C, "<span class='notice'>You feel less in-tune with your body.</span>")
+
+/obj/item/implant/health_monitor/implant(mob/living/target, mob/user, silent = 0)
```
Comment (Cyberboss) [inline code/game/objects/items/implants/implant_misc.dm]: FALSE

File: code/game/objects/items/implants/implant.dm [hunk 17]
```diff
@@ -17,6 +17,9 @@
 /obj/item/implant/proc/on_death(emote, mob/living/carbon/source)
 	return
 
+/obj/item/implant/proc/on_life(mob/living/carbon/source)
+	return
```
Comment (Cyberboss) [inline code/game/objects/items/implants/implant.dm]: It'd be more performant to just make it use `process()` imo. @vuonojenmustaturska thought?

File: code/modules/mob/living/carbon/life.dm [hunk 18]
```diff
@@ -456,6 +461,35 @@
...
+		hud_used.healths.invisibility = INVISIBILITY_ABSTRACT
+	for(var/X in hud_used.adv_health)
+		var/obj/screen/adv_health/S = X
+		if(adv_health_hud)
+			S.invisibility = 0
+			var/dmg_amt = get_damage_amount(S.dmg_type)
+			var/g = LERP(200, 0, CLAMP(dmg_amt, 0, 75)/75)
+			var/r = -g*0.11+95
+			if(stat == DEAD)
+				S.icon_state = "ded"
```
Comment (Cyberboss) [inline code/modules/mob/living/carbon/life.dm]: All these invisibility and icon_state changes belong in `update_icon()`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/__DEFINES/maths.dm (hunk 11 above)
Comment (MoreRobustThanYou) [reply code/__DEFINES/maths.dm]: this was changed when LERP was super bugged, i'll change it back

**Comment by fludd12** [comment]
I don't mean to, like, like holes or anything, but telling the difference between the dark green and orange heart rate thing is already pretty hard for me- it's almost impossible to tell with these sprites for me.

I totally get if you don't change it though, it's like, 0.5% of the population who has red/green issues, and it doesn't come up often in game anyway

**Comment by MoreRobustThanYou** [comment]
@fludd12 tbh i don't think the colors are perfect but that was the best i could get

**Comment by nicbn** [comment]
Just click yourself with help intent

**Comment by KorPhaeron** [comment]
Keeps evading and triggering the stickyban on server and causing a big mess with false positives and his PRs are the only thing he might actually care enough about to get him to stop

So his PRs are closed

**Comment by nicbn** [comment]
Beep bop bees server and codebase are separate things

**Comment by AlexTheSysop42** [comment]
>stickyban is pretty broken
>close his PRs because someone else's code is broken

Unless he's doing that specifically to mess other people up, this is dumb

**Comment by GunHog** [comment]
What Kor means to say is that MRTY himself is banned, and is trying to evade his ban. The stickyban system detects this and applies bans to cover any alternate IP addresses he used during his attempt. If he uses an IP address shared by other people (Such as a public Wi-Fi or proxy server), they would be affected as well.

**Comment by AlexTheSysop42** [comment]
Yeah I know that but why close his PRs because he's serverbanned?

Also, how did the mothsuperhero stickyban issue affect everyone, including people not using public IPs?

**Comment by Bawhoppen** [comment]
Close his PRs to get his attention so he stops messing with trying to evade.

**Comment by Bawhoppen** [comment]
No those posts were from a hacker and the PRs being closed were wholly unrelated

**Comment by MoreRobustThanYou** [comment]
Hahahaha. Anyways, I didn't trigger the stickyban, apparently it's being investigated

**Comment by nicbn** [comment]
Anyway that is not good. You can just help intent yourself.

**Comment by MoreRobustThanYou** [comment]
@nicbn that doesn't show toxin/oxy damage.

**Comment by Iamgoofball** [comment]
@vuonojenmustaturska don't let him keep opening PRs

**Comment by AlexTheSysop42** [comment]
why

**Comment by Iamgoofball** [comment]
mrty pr

**Pull request closed without merging**